website/blog/deployers/multi/main.go

```diff
@@ -0,0 +1,144 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+
+	"github.com/ServiceWeaver/weaver/runtime/colors"
+	"github.com/ServiceWeaver/weaver/runtime/envelope"
+	"github.com/ServiceWeaver/weaver/runtime/logging"
+	"github.com/ServiceWeaver/weaver/runtime/protos"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+// deployer is a simple multiprocess deployer that doesn't implement
+// co-location or replication. That is, every component is run in its own OS
+// process, and there is only one replica of every component.
+type deployer struct {
+	mu        sync.Mutex           // guards weavelets
+	weavelets map[string]*weavelet // weavelets, by component
+}
+
+// weavelet contains information about a weavelet.
+type weavelet struct {
+	envelope *envelope.Envelope // envelope to the weavelet
+	address  string             // weavelet's address
+}
+
+// handler handles messages from a weavelet.
+type handler struct {
+	deployer *deployer          // underlying deployer
+	envelope *envelope.Envelope // envelope to the weavelet
+}
+
+// Check that handler implements the envelope.EnvelopeHandler interface.
+var _ envelope.EnvelopeHandler = &handler{}
+
+// The unique id of the application deployment.
+var deploymentId = uuid.New().String()
+
+func main() {
+	flag.Parse()
+	d := &deployer{weavelets: map[string]*weavelet{}}
+	d.spawn("main") //nolint:errcheck // omitted for brevity
+	select {}       // block forever
+}
+
+// spawn spawns the provided component (if it hasn't already spawned) and
+// returns the spawned weavelet hosting the component.
+func (d *deployer) spawn(component string) (*weavelet, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Check if the component has already been spawned.
+	w, ok := d.weavelets[component]
+	if ok {
+		// The component has already been spawned.
+		return w, nil
+	}
+
+	// Spawn a weavelet to host the component.
+	info := &protos.EnvelopeInfo{
+		App:           "app",               // the application name
+		DeploymentId:  deploymentId,        // the deployment id
+		Id:            uuid.New().String(), // the weavelet id
+		SingleProcess: false,               // is the app a single process?
+		SingleMachine: true,                // is the app on a single machine?
+		RunMain:       component == "main", // should the weavelet run main?
+	}
+	config := &protos.AppConfig{
+		Name:   "app",       // the application name
+		Binary: flag.Arg(0), // the application binary
+	}
+	envelope, err := envelope.NewEnvelope(context.Background(), info, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Inform the weavelet of the component it should host.
+	if err := envelope.UpdateComponents([]string{component}); err != nil {
```

Comment:
I have a marginal preference to keeping it outside the mutex just because in general it isn't great to make RPCs while holding mutex (distributed deadlock possibilities are hard to catch and debug).